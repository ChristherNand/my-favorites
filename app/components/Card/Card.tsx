import React, { Dispatch, useRef } from 'react';
import './Card.css';

import Icon from '../generals/Icon';
import Tag from '../generals/Tag';
import Tooltip from '../generals/Tooltip';
import { removeTagLinkAsyncAction, removeLinkAsyncAction } from '../../state/actions';
import { iLink } from '../../interfaces';
import { useAppDispatch } from '../../hooks/redux';

const fakeText = `Donec in venenatis metus. Suspendisse potenti. Cras ultricies turpis sit amet massa suscipit, in egestas sapien finibus.`;

const Card: React.FC<iLink> = ({ id = '0', title, url, tags = [], description = fakeText }) => {
  const containerUrl = useRef<HTMLButtonElement>(null);
  const [isContHover, setIsContHover] = React.useState(false);

  const dispatchRemoveLink: Dispatch<any> = useAppDispatch();
  const dispatchRemoveTag: Dispatch<any> = useAppDispatch();
  const removeLink = () => dispatchRemoveLink(removeLinkAsyncAction(id as string));
  const removeTag = (tagId: string) => dispatchRemoveTag(removeTagLinkAsyncAction(id as string, tagId));

  const editLink = () => {
    // Todo edit link method backed
    console.log('editLink...', id);
  }

  return (
    <div className="card">
      <a className="linkTitle" href={url} target="_blank" data-test="link-url" rel="noreferrer">
        <h2 className="title" data-test="title">
          {title}
        </h2>
      </a>

      <div className="contDescription">
        <button
          className="description"
          ref={containerUrl}
          onMouseEnter={() => setIsContHover(true)}
          onMouseLeave={() => setIsContHover(false)}
          data-test="description"
        >
          {description}
        </button>
        <Tooltip
          text={description}
          calcHeight={true}
          hover={isContHover}
          parentRef={containerUrl}
          data-test="cp-tooltip"
        />
      </div>

      <div className="card__cont contTags --flex --wrap">
        {tags.map((tag, index) => (
          <Tag
            data-test="cp-tag"
            key={`${id}-${index}`}
            isUpdateDisable={true}
            onClose={() => removeTag(tag.id as string)}
            {...tag}
          />
        ))}
      </div>
      <div className="card_options --flex">
        <div className="loading">
          <Icon
            name="edit"
            onClick={editLink}
            className={'edit iconHover'}
            pointer={true}
            data-test="btn-edit"
          />
        </div>
        <div className="loading">
          <Icon
            name="close"
            onClick={removeLink}
            className={'close iconHover'}
            pointer={true}
            data-test="btn-remove"
          />
        </div>
      </div>
    </div>
  );
}

export default Card;
