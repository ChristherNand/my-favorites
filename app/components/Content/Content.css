.Content{
  width: calc(100% - var(--sidebar-width));
  box-sizing: border-box;
  padding: 20px 20px 20px 0px;
  margin-left: var(--sidebar-width);
}
