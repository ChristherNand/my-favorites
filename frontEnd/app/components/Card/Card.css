.card{
  --num-cards: 3;
  --icon-size: 18px;
  --cards-margin: 10px;
  --elemts-margin-top: 10px;

  display: flex;
  padding: 10px;
  font-size: 16px;
  color: var(--gray);
  align-items: center;
  box-sizing: border-box;
  flex-direction: column;
  justify-content: center;
  background: var(--white);
  margin: calc(var(--cards-margin) / 2);
  width: calc(100% / var(--num-cards) - var(--cards-margin));
}

.card div:empty{
  display: none;
}

.linkTitle{
  width: 100%;
  text-decoration: none;
}

.description{
  font-size: 14px;
  max-height: 88px;
  white-space: pre-line;
}

.description, .card_options, .contTags{
  margin-top: var(--elemts-margin-top);
}

.contTags{
  width: 100%;
}

.material-icons.iconHover{
  opacity: 0.8;
  color: var(--blue);
  font-size: var(--icon-size);
}

.iconHover:hover{
  opacity: 1;
}

.loading{
  width: 24px;
  margin: 6px;
  height: 24px;
  display: flex;
  position: relative;
  justify-content: center;
  align-items: center;
}

.loading::before{
  content: '';
  width: 28px;
  height: 28px;
  position: absolute;
  border-radius: 50%;
  pointer-events: none;
  border: 4px solid var(--light-gray-4);
}

.loading::after{
  content: "";
  position: absolute;
  top: 48%;
  left: -6px;
  width: 4px;
  height: 4px;
  margin: -2px 0 0 0;
  border-radius: 50%;
  background: var(--gray);
  transform-origin: 18px 50%;
  animation-duration: 5s, 0.4s;
  animation-fill-mode: forwards;
  animation-iteration-count: 1, infinite;
  /* animation-iteration-count: infinite; */
  animation-timing-function: ease, linear;
  /* animation-name: anim-effect-simo-3; */
  animation-name: anim-effect-simo-2, anim-effect-simo-3;
}

@keyframes anim-effect-simo-2 {
	0%, 5%, 95%, 100% {
		opacity: 0;
	}
	10%, 90% {
		opacity: 1;
	}
}

@keyframes anim-effect-simo-3 {
	100% {
		-webkit-transform: rotate3d(0, 0, 1, 360deg);
		transform: rotate3d(0, 0, 1, 360deg);
	}
}

