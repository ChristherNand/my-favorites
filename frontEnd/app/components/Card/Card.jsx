import React, {useRef, useState} from 'react';
import {connect} from 'react-redux';
import {CardPropType} from '../../propsTypes';
import './Card.css';

import Icon from '../generals/Icon';
import Tag from '../generals/Tag';
import Tooltip from '../generals/Tooltip';
import {removeTagLink} from "../../state/actions";

const mapDispachToProps = dispatch => ({
    removeTagLink: (linkId, tagId) => dispatch(removeTagLink(linkId, tagId)),
  })

const fakeText = `Donec in venenatis metus. Suspendisse potenti. Cras ultricies turpis sit amet massa suscipit, in egestas sapien finibus. 
  Quisque id massa pretium felis auctor mattis. Donec elementum est ut nisi finibus auctor.`;

function Card(props) {
  const {id = 0, title, url, tags = [], description = fakeText} = props;
  const containerUrl = useRef(null);
  const [isContHover, setIsContHover] = useState(false);

  function removeTag(tagId) {
    props.removeTagLink(props.id, tagId);
  }
  
  function removeCard() {
    console.log("cardId", props.id)
  }

  return  <div className="card">
            <a className="linkTitle" href={url} target="_blank">
              <h2 className="title">{title}</h2>
            </a>

            <div className="contDescription">
              <p className="description" ref={containerUrl} onMouseEnter={() => setIsContHover(true)} onMouseLeave={() => setIsContHover(false)}>
                {description}
              </p>
              <Tooltip text={description} calcHeight={true} hover={isContHover} parentRef={containerUrl}/>
            </div>

            <div className="card__cont contTags --flex --wrap">{
              tags.map((tag, index) => {
                return <Tag key={`${id}-${index}`}
                            updateDisable={true} {...tag}
                            onClose={() => removeTag(tag.id)}
                        />
              })}
            </div>
            <div className="card_options --flex">
              <div className="loading">
                <Icon name="edit" onClick={removeCard} className={"edit iconHover"} pointer={true}/>
              </div>
              <div className="loading">
                <Icon name="close" onClick={removeCard} className={"close iconHover"} pointer={true}/>
              </div>
            </div>
          </div>
}

Card.propTypes = CardPropType

export default connect(null, mapDispachToProps)(Card);
