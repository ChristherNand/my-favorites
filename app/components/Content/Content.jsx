import React, {useEffect} from 'react';
import {connect} from "react-redux";
import PropTypes from "prop-types";
import {CardPropType} from "../../propsTypes";
import "./Content.css";

import Card from "../Card";
import Search from "../Search";
import Row from "../generals/Row";
import CardLoading from "../CardLoading";
import TagList from "../generals/TagList";

import {getAllLinks, searchLinkAction} from "../../state/actions";

const mapStateToProps = state => ({
  tags: state.tags,
  links: state.links,
})

const mapDispachToProps = dispatch => ({
        getAllLinks: () => dispatch(getAllLinks()),
        searchLink: text => dispatch(searchLinkAction(text)),
      })
      
export function Content(props) {
  const {links = [], numLoadingCards = 9, searchLink, getAllLinks, tags} = props;

  useEffect(() => {
    getAllLinks();
  }, [])

  const isLoadingLinks = links && links.length == 0;
  function cardsLoading(){
    return Array.from({length: numLoadingCards})
              .map((_, index) => <CardLoading key={`cardLoading-${index}`}/>)
  }

  function cardsLoaded(links){
    return links.map((info, index) => <Card key={`card-${index}`} {...info}/>)
  }

  return  <section className="Content">
            <Row className="--flexColumn">
              <Search links={links} searchLink={searchLink} getAllLinks={getAllLinks}/>
              <TagList
                tags={tags} 
                autoHide={false}
                className="tagList"
                propertyFilter="name"
                clearAfterSelecting={true}
                placeHolder="Filter By Tags"
                onTagsSaved={tags => console.log(tags)}
              />
            </Row>
            <Row className="--wrap --spaceEvenly">
              {isLoadingLinks? cardsLoading(): cardsLoaded(links)}
            </Row>
          </section>
}

Content.propTypes = {
    getAllLinks: PropTypes.func,
    links: PropTypes.arrayOf(PropTypes.shape(CardPropType)),
}

export default connect(mapStateToProps, mapDispachToProps)(Content);
