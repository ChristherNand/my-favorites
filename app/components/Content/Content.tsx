import React, { Dispatch, useEffect } from 'react';
import './Content.css';

import { iContent, iLink, iTag } from '../../interfaces';

import { useAppDispatch, useAppSelector } from '../../hooks/redux';
import { RootState } from '../../state/store';
import { getAllLinksAction, getAllLinksByTagsAction, searchLinkAction } from '../../state/actions';
import CardLoading from '../CardLoading';
import Card from '../Card';
import Row from '../generals/Row';
import Search from '../Search';
import TagList from '../generals/TagList';

export enum eContent {
  tagListClass = 'tagList',
  tagListPlaceHolder = 'Filter By Tags',
}

const selectLinks = (state: RootState) => state.links;
const selectTags = (state: RootState) => state.tags;

const Content: React.FC<iContent> = ({ numLoadingCards = 2 }) => {
  const links = useAppSelector(selectLinks);
  const tags = useAppSelector(selectTags);
  const isLoadingLinks = links?.length == 0;

  const dispatchGetAllLinks: Dispatch<any> = useAppDispatch();
  const dispatchGetAllLinksByTags: Dispatch<any> = useAppDispatch();
  const dispatchSearchLink = useAppDispatch();

  const getAllLinks = () => dispatchGetAllLinks(getAllLinksAction());
  const searchLink = (text: string) => dispatchSearchLink(searchLinkAction(text));
  const getAllLinksByTags = (tagsIds: string) =>
    dispatchGetAllLinksByTags(getAllLinksByTagsAction(tagsIds));

  useEffect(() => {
    getAllLinks();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const cardsLoading = () =>
    Array.from({ length: numLoadingCards }, (_, index) => (
      <CardLoading key={`cardLoading-${index}`} data-test="cp-cardLoading" />
    ));

  const cardsLoaded = (links: iLink[]) =>
    links.map((info, index) => <Card key={`card-${index}`} {...info} data-test="cp-card" />);

  const getTagsIds = (selectedTags: iTag[] = []) =>
    selectedTags.reduce((accum, { id }): string => {
      return accum === '' ? (id as string) : `${accum},${id}`;
    }, '');

  const requestLinksByTags = (selectedTags: iTag[]) => getAllLinksByTags(getTagsIds(selectedTags));

  return (
    <section className="Content">
      <Row className="--flexColumn">
        <Search onSearchLink={searchLink} onGetAllLinks={getAllLinks} data-test="cp-search" />
        <TagList
          options={tags}
          autoHide={false}
          data-test="cp-tagList"
          clearAfterSelecting={true}
          className={eContent.tagListClass}
          placeHolder={eContent.tagListPlaceHolder}
          onTagsSaved={requestLinksByTags}
        />
      </Row>
      <Row className="--wrap --spaceEvenly">
        {isLoadingLinks ? cardsLoading() : cardsLoaded(links)}
      </Row>
    </section>
  );
};

export default Content;
