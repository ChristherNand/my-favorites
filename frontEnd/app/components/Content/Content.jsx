import React, {useEffect} from 'react';
import {connect} from "react-redux";
import PropTypes from "prop-types";
import {LinkPropType} from "../../propsTypes";
import "./Content.css";

import Link from "../Link";
import Search from "../Search";
import Row from "../generals/Row";
import CreateLink from "../CreateLink";

import {getAllLinks} from "../../state/actions";

const mapStateToProps = state => {
        return {links: state.links}
      }

const mapDispachToProps = dispatch => {
        return {
          getAllLinks: () => dispatch(getAllLinks()),
        }
      }

function Content(props) {
  useEffect(() => {
      props.getAllLinks();
  }, [])

  return  <section className="Content">                
            <Row className="--flexEnd"><Search style={{"width": "30%"}}/></Row>
            <Row><CreateLink/></Row>

            {props.links.map((info, index) => <Row key={`${info.date}-${index}`}>
                                                <Link {...info}/>
                                              </Row>)}
          </section>
}

Content.propTypes = {
    getAllLinks: PropTypes.func,
    links: PropTypes.arrayOf(PropTypes.shape(LinkPropType)),
}

export default connect(mapStateToProps, mapDispachToProps)(Content);
