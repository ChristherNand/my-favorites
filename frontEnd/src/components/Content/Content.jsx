import React, {useEffect} from 'react';
import {connect} from "react-redux";
import "./Content.css";

import Link from "../Link";
import Search from "../Search";
import Row from "../generals/Row";
import CreateLink from "../CreateLink";

import {useStateValueCtx} from "../../context/Tags.contex";
import {getLinks} from "../../utils/ServerRequest";

const mapStateToProps = state => {
    return {links: state.links}
}

function Content(props) {

    /* const [dataCtx, dispatchCtx] = useStateValueCtx(); */

    useEffect(() => {
        //requestLinks();
        console.log(props);
    }, [])
    
    async function requestLinks(){
        let links = await getLinks();
        /* dispatchCtx({type: "addLinks", links}); */
    }

    return  <section className="Content">                
                <div className="contend">
                    <Row className="--flexEnd">{() => <Search style={{"width": "30%"}}/>}</Row>
                    <Row>{() => <CreateLink />}</Row>

                    {props.links.map((info, index) =>  <Row key={`${info.date}-${index}`}>
                                                    {() => <Link {...info}/>}
                                                </Row>)}
                </div>
            </section>
}

export default connect(mapStateToProps)(Content);
