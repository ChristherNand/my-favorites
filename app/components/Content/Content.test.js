import React from 'react';
import { shallow, mounted } from 'enzyme';

import { Content } from './Content';
import Card from '../Card';

var Component;
const props = {
        links: [],
        numLoadingCards: 10,
        getAllLinks: jest.fn()
      }
const linkData = {id: 'id test', title: 'title test', url: 'test url', description: 'test description'}

beforeEach(() => {
  Component = shallow(<Content {...props}/>);    
});

test('take snapshot', () => {
  expect(Component).toMatchSnapshot();
});

test('render loading links', () => {
  expect(Component.find('CardLoading')).toHaveLength(props.numLoadingCards);
});

test('render loading links', () => {
  expect(Component.find('CardLoading')).toHaveLength(props.numLoadingCards);
});

test.skip('getAllLinks is called', () => {
  Component = shallow(<Content {...props}/>);    
  expect(props.getAllLinks).toHaveBeenCalled();
});

test('render snapshot of final links', () => {
  Component = shallow(<Content {...props} links={[linkData, linkData, linkData]}/>);
  expect(Component).toMatchSnapshot();
});

test('render final links', () => {
  Component = shallow(<Content {...props} links={[linkData, linkData, linkData]}/>);
  expect(Component.find(Card)).toHaveLength(3);
  expect(Component.find(Card).at(0).props()).toEqual(linkData);
});


