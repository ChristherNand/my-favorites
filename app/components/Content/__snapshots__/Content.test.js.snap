// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render snapshot of final links 1`] = `
<section
  className="Content"
>
  <Row
    className="--flexEnd"
  >
    <Search
      getAllLinks={[MockFunction]}
      links={
        Array [
          Object {
            "description": "test description",
            "id": "id test",
            "title": "title test",
            "url": "test url",
          },
          Object {
            "description": "test description",
            "id": "id test",
            "title": "title test",
            "url": "test url",
          },
          Object {
            "description": "test description",
            "id": "id test",
            "title": "title test",
            "url": "test url",
          },
        ]
      }
    />
  </Row>
  <Row
    className="--wrap --spaceEvenly"
  >
    <Connect(Card)
      description="test description"
      id="id test"
      key="card-0"
      title="title test"
      url="test url"
    />
    <Connect(Card)
      description="test description"
      id="id test"
      key="card-1"
      title="title test"
      url="test url"
    />
    <Connect(Card)
      description="test description"
      id="id test"
      key="card-2"
      title="title test"
      url="test url"
    />
  </Row>
</section>
`;

exports[`take snapshot 1`] = `
<section
  className="Content"
>
  <Row
    className="--flexEnd"
  >
    <Search
      getAllLinks={[MockFunction]}
      links={Array []}
    />
  </Row>
  <Row
    className="--wrap --spaceEvenly"
  >
    <CardLoading
      key="cardLoading-0"
    />
    <CardLoading
      key="cardLoading-1"
    />
    <CardLoading
      key="cardLoading-2"
    />
    <CardLoading
      key="cardLoading-3"
    />
    <CardLoading
      key="cardLoading-4"
    />
    <CardLoading
      key="cardLoading-5"
    />
    <CardLoading
      key="cardLoading-6"
    />
    <CardLoading
      key="cardLoading-7"
    />
    <CardLoading
      key="cardLoading-8"
    />
    <CardLoading
      key="cardLoading-9"
    />
  </Row>
</section>
`;
